Early logging improvements

ajtowns opened the pull request:
In order to cope gracefully with `Log*()` calls that are invoked prior to logging being fully configured (indicated by calling `StartLogging()` we buffer early log messages in `m_msgs_before_open`. This has a couple of minor issues:

 * if there are many such log messages the buffer can become arbitrarily large; this can be a problem for users of libkernel that might not wish to worry about logging at all, and as a result never invoke `StartLogging()`
 * early log messages are formatted before the formatting options are configured, leading to inconsistent output

Fix those issues by buffering the log info prior to formatting it, and setting a limit on the size of the buffer (dropping the oldest lines, and reporting the number of lines skipped).

Also adds some thread safety annotations, and the ability to invoke `LogInstance().DisableLogging()` if you want to disable logging entirely, for a minor efficiency improvement.

ajtowns committed:
logging: Add thread safety annotations

ajtowns committed:
logging: Limit early logging buffer

Log messages created prior to StartLogging() being called go into a
buffer. Enforce a limit on the size of this buffer.

ajtowns committed:
logging: Add DisableLogging()

ajtowns commented:
Easiest way to test is to add some log lines into init.cpp `InitLogging()`, eg:

```c++
    for (int i = 0; i < 10000; ++i) {
        LogInfo("log spam %d...\n", i);
    }
    LogWarning("awoogah\n");
    LogError("internal bug not detected\nhello, world!\nwas that fun?\n");
```

Running `bitcoind -regtest -logtimemicros` should give a decent indication of the differences.

If you want to modify bitcoin-chainstate to see what the logs would look like, remove the `DisableLogging` call and add:

```c++
LogInstance().m_print_to_console = true;
LogInstance().StartLogging()
```

somewhere.

ryanofsky commented on src/logging.cpp:
```diff
@@ -71,6 +72,9 @@ bool BCLog::Logger::StartLogging()
 
     // dump buffered messages from before we opened the log
     m_buffering = false;
+    if (m_buffer_lines_skipped > 0) {
```
In commit "logging: Limit early logging buffer" (3d4b1f4af406276671b5d0115574dea6b4a64340)

Since the log limit discards old log prints instead of skipping new prints, I think it would be a little clearer if "skipped" was replaced with "discarded" throughout this commit.

ajtowns replied:
Changed

ryanofsky commented on src/logging.cpp:
```diff
@@ -382,11 +402,20 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
 
     str_prefixed = LogTimestampStr(str_prefixed);
 
-    m_started_new_line = !str.empty() && str[str.size()-1] == '\n';
```
In commit "logging: Limit early logging buffer" (3d4b1f4af406276671b5d0115574dea6b4a64340)

It seems like a mistake to drop this as it is not being set otherwise.

ajtowns replied:
Should be fixed (line is no longer touched)

ryanofsky commented on src/logging.cpp:
```diff
@@ -372,9 +368,24 @@ std::string BCLog::Logger::GetLogPrefix(BCLog::LogFlags category, BCLog::Level l
+    return buflog.str.size() + buflog.logging_function.size() + buflog.source_file.size() + buflog.threadname.size() + memusage::MallocUsage(sizeof(memusage::list_node<BCLog::Logger::BufferedLog>));
+}
+
+void BCLog::Logger::FormatLogStrInPlace(std::string& str, BCLog::LogFlags category, BCLog::Level level, std::string source_file, int source_line, std::string logging_function, std::string threadname, SystemClock::time_point now, std::chrono::seconds mocktime) const
```
In commit "logging: Apply formatting to early log messages" (0456828a50b86fd9185b7428e9012b7c18e970f9)

Maybe something to consider for a followup but some of the arguments to FormatLogStrInPlace and the two LogPrintStr functions could be string_views instead of strings (logging_function, source_file, threadname).

ajtowns replied:
Added a followup commit to switch to string_view

ryanofsky reviewed (approved):
Code review ACK 0456828a50b86fd9185b7428e9012b7c18e970f9. Left some suggestions, but they are not important

ajtowns committed:
logging: Apply formatting to early log messages

The formatting of log messages isn't defined until StartLogging() is
called; so can't be correctly applied to early log messages from prior
to that call. Instead of saving the output log message, save the inputs
to the logging invocation (including time, mocktime and thread name),
and format those inputs into a log message when StartLogging() is called.

ajtowns committed:
logging: use std::string_view

ajtowns commented on src/bitcoin-chainstate.cpp:
```diff
@@ -42,6 +43,8 @@
 
 int main(int argc, char* argv[])
 {
+    LogInstance().DisableLogging();
```
It's mostly added so that `DisableLogging()` isn't dead code; all it really achieves is a minor performance improvement. Added some comments that maybe clarifies things.

ryanofsky commented on src/logging.h:
```diff
@@ -157,6 +157,14 @@ namespace BCLog {
+         * Mostly intended for libbitcoin-kernel apps that don't want any logging.
+         * Should be used instead of StartLogging().
+         */
+        void DisableLogging() EXCLUSIVE_LOCKS_REQUIRED(!m_cs);
```
In commit "logging: Add DisableLogging()" (0b1960f1b29cfe5209ac68102c8643fc9553f247)

I think it's a good idea to add this `DisableLogging()` method for clarity, even though in practice it's equivalent to just calling `StartLogging()` which setting any logging options.

But I think in the future it would be nice to simplify the log interface and replace `StartLogging()` `DisableLogging()` and `DisconnectTestLogger()` methods with a single SetOptions() method:

```c++
struct BufferOptions {
   size_t max_bytes{1000000};
};

struct OutputOptions {
    fs::path print_to_file{};
    bool print_to_console{false};
    bool log_timestamps{true};
    bool log_time_micros{false};
    bool log_threadnames{false};
    bool log_sourcelocations{false};
    bool always_print_category_level{false};
};

using Options = std::variant<BufferOptions, OutputOptions>;

Options m_options{};

void SetOptions(Options options);
```

The default behavior, as currently, would be to buffer log messages, but it would be easy to set default output options to disable logging, or set custom output options to enable it.

ajtowns replied:
I'd like to change `Enabled()` and `WillLogCategoryLevel()` functions to be lockless so that calls to `Log*()` don't block when they have nothing to do. I think the above changes would mesh well with that.

ryanofsky commented on src/logging.cpp:
```diff
@@ -387,40 +398,44 @@ void BCLog::Logger::LogPrintStr_(const std::string& str, const std::string& logg
-            if (m_msgs_before_open.empty()) {
-                m_cur_buffer_memusage = 0;
-                break;
+        if (!starts_new_line && !m_msgs_before_open.empty()) {
```
In commit "logging: Apply formatting to early log messages" (a67baea866719837865864a5eb56a47a9f132de1)

I think it might make sense to explicitly handle the case where `starts_new_line` is false and `m_msgs_before_open` is empty. Right now the code will cut off the first part of the log line but include the rest of the message, which could lead to confusing or misleading output. Better options might be to add a `... [truncated]` prefix to the line, or to discard the whole message.

ajtowns replied:
Added a `[...]` marker, since it was easy.

In order to trigger it you'd need to have 1MB of log message on a single, incomplete line, which seems a bit ridiculous. Long-term, I think it would make more sense to stop intentionally doing incomplete lines (since that introduces a race condition if other threads are also logging), and just add a newline if it's missing... *shrug*

ryanofsky commented on src/logging.cpp:
```diff
@@ -373,28 +373,37 @@ static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
 
     if (m_log_sourcelocations) {
-        str.insert(0, "[" + RemovePrefix(source_file, "./") + ":" + ToString(source_line) + "] [" + logging_function + "] ");
+        insert_many(str, 0, "[", RemovePrefix(source_file, "./"), ":", ToString(source_line), "] [", logging_function, "] ");
```
In commit "logging: use std::string_view" (f3632d53a5be7efe61be8181ff481d30f7313bd4)

I wonder if using insert_many is actually more efficient than strprinf:

```c++
str.insert(0, strprintf("[%s:%d] [%s] ", RemovePrefix(source_file, "./"), source_line, logging_function);
```

Using strprintf seems more readable, and I suspect if we really wanted to optimize this code we probably would build up a new string rather than repeatedly inserting at the beginning of an existing string.

ajtowns replied:
Changed to strprintf. Got a bit target fixated there...

ryanofsky reviewed (approved):
Code review ACK f3632d53a5be7efe61be8181ff481d30f7313bd4. I left some suggestions, but feel free to ignore them. Overall I think this PR makes early logging behavior a lot better by making the log format consistent, and it also makes nice API improvements like avoiding runaway memory usage, and switching to std::string_view.

maflcko reviewed (approved):
ACK f3632d53a5be7efe61be8181ff481d30f7313bd4 🚲

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK f3632d53a5be7efe61be8181ff481d30f7313bd4 🚲
Npt3Z0ZvZ7d1vkjP+I6WoSxmYlnR8J954hzJEEVA5Y1ogrEp3GDcMINyxhP04yvRf/5iWhedn+y9a9jxwtHmAw==
```

</details>

ajtowns commented on src/logging.cpp:
```diff
@@ -96,6 +96,18 @@ void BCLog::Logger::DisconnectTestLogger()
+    {
+        StdLockGuard scoped_lock(m_cs);
+        assert(m_buffering);
+        assert(m_print_callbacks.empty());
```
The other things can be safely set to false, so we can be sure they're false without an assertion. But if you've set a callback you've also gotten an `iterator` that you might intend to pass to `DeleteCallback` later, which would then crash if we had cleared `m_print_callbacks` here.

ajtowns commented on src/logging.cpp:
```diff
@@ -387,40 +398,44 @@ void BCLog::Logger::LogPrintStr_(const std::string& str, const std::string& logg
-                break;
+        if (!starts_new_line && !m_msgs_before_open.empty()) {
+            m_msgs_before_open.back().str += str;
+            m_cur_buffer_memusage += str.size();
```
That's a bug, changed.

ajtowns commented on src/logging.cpp:
```diff
@@ -387,40 +398,44 @@ void BCLog::Logger::LogPrintStr_(const std::string& str, const std::string& logg
+                .level=level,
+            };
+            m_msgs_before_open.push_back(buf);
+            m_cur_buffer_memusage += MemUsage(buf);
```
Done

ryanofsky commented on src/logging.cpp:
```diff
@@ -387,28 +398,37 @@ void BCLog::Logger::LogPrintStr_(const std::string& str, const std::string& logg
+                return;
+            } else {
+                // unlikely edge case; add a marker that something was trimmed
+                str_prefixed.insert(0, "[...] ");
```
In commit "logging: Apply formatting to early log messages" (558df5c733d31456faf856d44f7037f41981d797)

Probably better to leave alone since this affects an edge case of an edge case, but just observing that this code could increment m_cur_buffer_memusage to count the usage more accurately.

ajtowns replied:
The memusage is incremented later when `str_prefixed` is included in `buf`, so afaics this should already be accurate.

ryanofsky reviewed (approved):
Code review ACK b4dd7ab43e8cfc2c171f67588e4e1ec2705393c2

Main change since last review is fixing an escaping bug https://github.com/bitcoin/bitcoin/pull/30386#discussion_r1683742656, but also includes a few other review suggestions

sedited reviewed (approved):
Nice, ACK b4dd7ab43e8cfc2c171f67588e4e1ec2705393c2
